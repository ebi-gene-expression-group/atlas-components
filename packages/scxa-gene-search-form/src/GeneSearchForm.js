import React from 'react'
import PropTypes from 'prop-types'
import URI from 'urijs'

import Autocomplete from './Autocomplete'
import SpeciesSelect from './SpeciesSelect'

class GeneSearchForm extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      // A JSON-formatted query object with two fields, term and category
      query: {},
      selectedSpecies: props.defaultSpecies
    }

    this.speciesSelectOnChange = this._speciesSelectOnChange.bind(this)
    this.autocompleteOnChange = this._autocompleteOnChange.bind(this)
  }

  _autocompleteOnChange(selectedItem) {
    this.setState({
      query: selectedItem ?
        JSON.parse(selectedItem.value) :
        {}
    })
  }

  _speciesSelectOnChange(event) {
    this.setState({ selectedSpecies: event.target.value })
  }

  render() {
    const {wrapperClassName, actionEndpoint} = this.props

    const {autocompleteClassName, atlasUrl, suggesterEndpoint, defaultValue} = this.props

    const {enableSpeciesSelect, speciesSelectClassName, speciesSelectStatusMessage} = this.props
    const {allSpecies, topSpecies} = this.props

    return (
      <form action={URI(actionEndpoint, atlasUrl).toString()} method={`post`}>
        <div className={wrapperClassName}>
          <div className={autocompleteClassName}>
            <Autocomplete atlasUrl={atlasUrl}
                          suggesterEndpoint={suggesterEndpoint}
                          onChange={this.autocompleteOnChange}
                          selectedSpecies={this.state.selectedSpecies}
                          allSpecies={allSpecies}
                          defaultValue={defaultValue}/>
          </div>
          { enableSpeciesSelect &&
            <div className={speciesSelectClassName}>
              <SpeciesSelect allSpecies={allSpecies}
                             topSpecies={topSpecies}
                             statusMessage={speciesSelectStatusMessage}
                             selectedValue={this.state.selectedSpecies}
                             onChange={this.speciesSelectOnChange}/>
            </div>
          }
        </div>
        <div className={wrapperClassName}>
          <div className={`small-12 columns`}>
            <button type={`Submit`}
                    className={`button`}
                    disabled={!this.state.query.term || this.state.query.term.trim() === ``}>
                    Search
            </button>
          </div>
        </div>
      </form>
    )
  }
}

GeneSearchForm.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  actionEndpoint: PropTypes.string.isRequired,
  wrapperClassName: PropTypes.string,

  autocompleteClassName: PropTypes.string,
  suggesterEndpoint: PropTypes.string.isRequired,
  defaultValue: PropTypes.shape({
    term: PropTypes.string,
    category: PropTypes.string
  }),

  enableSpeciesSelect: PropTypes.bool,
  speciesSelectClassName: PropTypes.string,
  allSpecies: PropTypes.arrayOf(PropTypes.string),
  topSpecies: PropTypes.arrayOf(PropTypes.string),
  defaultSpecies: PropTypes.string,

  speciesSelectStatusMessage: PropTypes.string
}

GeneSearchForm.defaultProps = {
  wrapperClassName: ``,
  autocompleteClassName: ``,
  defaultValue: {},
  enableSpeciesSelect: false,
  speciesSelectClassName: ``,
  allSpecies: [],
  topSpecies: [],
  defaultSpecies: ``,
  speciesSelectStatusMessage: ``
}

export default GeneSearchForm
