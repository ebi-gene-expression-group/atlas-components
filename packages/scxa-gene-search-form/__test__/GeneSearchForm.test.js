import React from 'react'
import Enzyme from 'enzyme'
import {mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import GeneSearchForm from '../src/GeneSearchForm.js'
// import AsyncCreatableSelect from 'react-select/lib/AsyncCreatable'

Enzyme.configure({ adapter: new Adapter() })

const props = {
  atlasUrl: `foo/`,
  actionEndpoint: `bar`,
  suggesterEndpoint: `suggest`,
  enableSpeciesSelect: true
}

const defaultValue = {
  term: `foo`,
  category: `bar`
}

const species = [
  `Meeseek`,
  `Gromflomite`,
  `Cromulon`,
  `Zigerion`,
  `Moopian`,
  `Bliznarvian`,
  `Greebybobe`
]

describe(`GeneSearchForm`, () => {
  test(`search button is initially disabled`, () => {
    const wrapper = mount(<GeneSearchForm {...props}/>)
    expect(wrapper.find(`button`).at(0).props()).toHaveProperty(`disabled`, true)
  })

  test(`search button with invalid default value is disabled`, () => {
    const wrapper = mount(<GeneSearchForm {...props} defaultValue={{term: `   `, category: `q`}}/>)
    expect(wrapper.find(`button`).at(0).props()).toHaveProperty(`disabled`, true)
  })

  test(`search button with valid default value is enabled`, () => {
    const wrapper = mount(<GeneSearchForm {...props} defaultValue={defaultValue}/>)
    expect(wrapper.find(`button`).at(0).props()).toHaveProperty(`disabled`, false)
  })

})
