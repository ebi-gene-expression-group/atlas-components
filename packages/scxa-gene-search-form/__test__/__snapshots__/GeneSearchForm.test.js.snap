// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneSearchForm matches snapshot 1`] = `
<form
  action="foo/bar"
  method="post"
>
  <div
    className=""
  >
    <div
      className=""
    >
      <div>
        <label
          htmlFor="geneQuery"
        >
          Gene ID or gene symbol
        </label>
        ,
        <div
          className="css-10nd86i"
          onKeyDown={[Function]}
        >
          <div
            className="css-12emboz"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3"
            >
              <div
                className="css-1492t68"
              >
                
              </div>
              <div
                className="css-1g6gooi"
              >
                <div
                  className=""
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <input
                    aria-autocomplete="list"
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    disabled={false}
                    id="react-select-8-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    spellCheck="false"
                    style={
                      Object {
                        "background": 0,
                        "border": 0,
                        "boxSizing": "content-box",
                        "color": "inherit",
                        "fontSize": "inherit",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "1px",
                      }
                    }
                    tabIndex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "overflow": "scroll",
                        "position": "absolute",
                        "top": 0,
                        "visibility": "hidden",
                        "whiteSpace": "pre",
                      }
                    }
                  >
                    
                  </div>
                </div>
              </div>
            </div>
            <div
              className="css-1wy0on6"
            />
          </div>
          <input
            name="geneQuery"
            type="hidden"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className=""
    >
      <div>
        <label
          htmlFor="species"
        >
          Species
        </label>
        <div
          className="css-10nd86i"
          onKeyDown={[Function]}
        >
          <div
            className="css-12emboz"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3"
            >
              <div
                className="css-xp4uvy"
              >
                Any
              </div>
              <input
                className="css-14uuagi"
                id="react-select-9-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={true}
                tabIndex="0"
                value=""
              />
            </div>
            <div
              className="css-1wy0on6"
            >
              <span
                className="sc-bdVaJa jcfyVD"
                cx={[Function]}
              />
            </div>
          </div>
          <input
            name="species"
            type="hidden"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className=""
  >
    <div
      className="small-12 columns"
    >
      <button
        className="button"
        disabled={true}
        type="Submit"
      >
        Search
      </button>
    </div>
  </div>
</form>
`;
